hibernate-core/src/main/java/org/hibernate/internal/FastSessionServices.java

```diff
@@ -0,0 +1,280 @@
+		this.defaultCacheRetrieveMode = determineCacheRetrieveMode( defaultSessionProperties );
+		this.initialSessionCacheMode = CacheModeHelper.interpretCacheMode( defaultCacheStoreMode, defaultCacheRetrieveMode );
+		this.discardOnClose = sessionFactoryOptions.isReleaseResourcesOnCloseEnabled();
+		this.defaultJdbcObservers = Collections.singletonList( new ConnectionObserverStatsBridge( sf ) );
+		this.defaultSessionEventListeners = sessionFactoryOptions.getBaselineSessionEventsListenerBuilder();
+		this.defaultLockOptions = initializeDefaultLockOptions( defaultSessionProperties );
+	}
+
+	private static LockOptions initializeDefaultLockOptions(final Map<String, Object> defaultSessionProperties) {
+		LockOptions def = new LockOptions();
+		LockOptionsHelper.applyPropertiesToLockOptions( defaultSessionProperties, () -> def );
+		return def;
+	}
+
+	private static <T> EventListenerGroup<T> listeners(EventListenerRegistry elr, EventType<T> type) {
+		return elr.getEventListenerGroup( type );
+	}
+
+	SqlTypeDescriptor remapSqlTypeDescriptor(SqlTypeDescriptor sqlTypeDescriptor) {
+		if ( !sqlTypeDescriptor.canBeRemapped() ) {
+			return sqlTypeDescriptor;
+		}
+
+		final SqlTypeDescriptor remapped = dialect.remapSqlTypeDescriptor( sqlTypeDescriptor );
+		return remapped == null ? sqlTypeDescriptor : remapped;
+	}
+
+	private static boolean isTransactionAccessible(SessionFactoryImpl sf, TransactionCoordinatorBuilder transactionCoordinatorBuilder) {
+		// JPA requires that access not be provided to the transaction when using JTA.
+		// This is overridden when SessionFactoryOptions isJtaTransactionAccessEnabled() is true.
+		if ( sf.getSessionFactoryOptions().getJpaCompliance().isJpaTransactionComplianceEnabled() &&
+				transactionCoordinatorBuilder.isJta() &&
+				!sf.getSessionFactoryOptions().isJtaTransactionAccessEnabled() ) {
+			return false;
+		}
+		return true;
+	}
+
+	private static Map<String, Object> initializeDefaultSessionProperties(SessionFactoryImpl sf) {
+		HashMap<String,Object> p = new HashMap<>();
+
+		//Static defaults:
+		p.putIfAbsent( AvailableSettings.FLUSH_MODE, FlushMode.AUTO.name() );
+		p.putIfAbsent( JPA_LOCK_SCOPE, PessimisticLockScope.EXTENDED.name() );
+		p.putIfAbsent( JPA_LOCK_TIMEOUT, LockOptions.WAIT_FOREVER );
+		p.putIfAbsent( JPA_SHARED_CACHE_RETRIEVE_MODE, CacheModeHelper.DEFAULT_RETRIEVE_MODE );
+		p.putIfAbsent( JPA_SHARED_CACHE_STORE_MODE, CacheModeHelper.DEFAULT_STORE_MODE );
+
+		//Defaults defined by SessionFactory configuration:
+		final String[] ENTITY_MANAGER_SPECIFIC_PROPERTIES = {
+				JPA_LOCK_SCOPE,
+				JPA_LOCK_TIMEOUT,
+				AvailableSettings.FLUSH_MODE,
+				JPA_SHARED_CACHE_RETRIEVE_MODE,
+				JPA_SHARED_CACHE_STORE_MODE,
+				QueryHints.SPEC_HINT_TIMEOUT
+		};
+		final Map<String, Object> properties = sf.getProperties();
+		for ( String key : ENTITY_MANAGER_SPECIFIC_PROPERTIES ) {
+			if ( properties.containsKey( key ) ) {
+				p.put( key, properties.get( key ) );
```

Comment:
Seems we can improve performance by using 'get()' result to avoid 'containsKey' overhead.
Also, we might consider 'getOrDefault' usage, so we might avoid some unnecessary default value assignment previously.

Reply:
Hi Nathan, these commits have been merged long ago.

If you want to improve code of Hibernate, please send a pull request with further changes associated to a JIRA - but also maybe choose wisely what to work on, your time is precious and limited and there's likely more important things to fix.

In particular the "get" is not safe to refactor as some properties might be set intentionally to `null`.

Reply:
thanks for responding. I knew the PR has been merged long time ago, just left some comments to communicate. Yeah, null value in map is a good point. Yeah, your team members' time is pretty limited, but contributors' are not, though reviewing contributor's PR takes team member's cycles. Thanks.